src/state.py

```diff
@@ -0,0 +1,155 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The charm state and configurations."""
+
+import enum
+import json
+import logging
+import re
+import typing
+
+import ops
+import pydantic
+
+from errors import ConfigurationError
+
+logger = logging.getLogger(__name__)
+
+HOSTNAME_CONFIG_NAME = "hostname"
+PATH_CONFIG_NAME = "path"
+BACKENDS_CONFIG_NAME = "backends"
+PROTOCOL_CONFIG_NAME = "protocol"
+
+
+class Protocol(str, enum.Enum):
+    """Protocol to request backends.
+
+    Attributes:
+        HTTP: Use HTTP for requests.
+        HTTPS: Use HTTPS for requests.
+    """
+
+    HTTP = "http"
+    HTTPS = "https"
+
+
+class Configuration(pydantic.BaseModel):
+    """Represents the configuration.
+
+    Attributes:
+        hostname: The hostname for the virtual host for this set of configuration.
+        path: The path for this set of configuration.
+        backends: The backends for this set of configuration.
+        protocol: The protocol to request the backends with. Can be http or
+            https.
+    """
+
+    hostname: typing.Annotated[str, pydantic.StringConstraints(min_length=1)]
+    path: typing.Annotated[str, pydantic.StringConstraints(min_length=1)]
+    backends: tuple[pydantic.IPvAnyAddress, ...]
+    protocol: Protocol
+
+    @pydantic.field_validator("hostname")
+    @classmethod
+    def validate_hostname(cls, value: str) -> str:
+        """Validate the hostname.
+
+        Args:
+            value: The value to validate.
+
+        Raises:
+            ValueError: The validation failed.
+
+        Returns:
+            The value after validation.
+        """
+        if len(value) > 255:
+            raise ValueError("Hostname cannot be longer than 255")
+
+        valid_segment = re.compile(r"(?!-)[A-Z\d-]{1,63}(?<!-)$", re.IGNORECASE)
```

Comment:
This might be a bit restrictive because only ASCII characters are allowed. But should be fine for a first implementation